### PR TITLE
Memprof sampling for blocks in the minor heap, allocated by C code.

### PR DESCRIPTION
**Pull request by jhjourdan**
Allocations ignored by this version
- Marshalling
- In the minor heap by natively-compiled OCaml code

Allocations potentially sampled
- In the major heap
- In the minor heap by C code and OCaml code in bytecode mode

### CODE REVIEW

File: runtime/memprof.c
Comment (gasche) [inline runtime/memprof.c]: called

File: runtime/memprof.c
Comment (gasche) [inline runtime/memprof.c]: the part

File: runtime/memprof.c
Comment (gasche) [inline runtime/memprof.c]: I don't understand why it's the right thing to tie this renewing behavior to `set_suspended` in this way. A comment in `caml_memprof_track_young` suggests that this behavior is desirable when re-enabling sampling at the end of the function, and that renewing more than necessary is safe. But:
- is it ever useful to renew minor sampling when suspending (when the argument is `1`?); if not, why not renew only on `0`?
- is it expected that the disable/re-enable bracketing in `caml_memprof_track_alloc_shr` will also renew minor sampling? (My guess is "yes" but I'm not sure.)

From a distance it looks like the intent would be clearer without this renewing in the `set_suspend` function, and explicit `renew` calls in the right places instead. (Or maybe two functions `caml_memprof_pause_sampling()`, `caml_memprof_resume_sampling()`, the latter renewing.)

File: runtime/memprof.c
Comment (gasche) [inline runtime/memprof.c]: Note: the largest value is picked because `caml_young_ptr` is decreasing.

File: runtime/caml/memory.h
Comment (gasche) [inline runtime/caml/memory.h:220]: Why did the previous test in the existing code use `caml_young_ptr < caml_young_trigger` instead of comparing to `caml_young_limit`? If I understand correctly, the change is safe because it will only dispatch more often, which is the right thing to do in the case where `caml_young_limit` is *not* `caml_young_trigger` (because it means that a signal is pending or a GC round has been requested).

File: runtime/memprof.c
Comment (jhjourdan) [reply runtime/memprof.c]: There are two ways you can imagine for suspending/resuming sampling:
1- (virtually) Setting lambda to 0
2- Ignoring all the samples that we generate

What I implemented is 1-: This means that, when suspended, everything happens as if `lambda=0`, i.e., `caml_memprof_young_trigger` becomes equal to `caml_young_alloc_start`, and `caml_memprof_track_young` is never entered. But then suspending/resuming involve  changing the value of lambda, which requires renewing the next minor sample. Note, however,  that renewing in the suspended state is very cheap (no random sampling is performed), so the actual renewing is only performed once.

The other possibility would be to implement 2-. I don't think the performance would be significantly different, but probably the code is going to be simpler. Let me try.

File: runtime/memprof.c
Comment (jhjourdan) [reply runtime/memprof.c]: Yes. Do you really think this is worth a comment?

File: runtime/caml/memory.h
Comment (jhjourdan) [reply runtime/caml/memory.h:220]: In bytecode mode, the two variables used to be equal.

In native code mode, you are almost right: this is correct because we would only dispatch more often. There is a subtlety, though, because `caml_alloc_small_dispatch` do not check for signals, nor for `caml_requested_major_slice`/`caml_requested_minor_gc`. I'll add that.

File: runtime/memprof.c
Comment (gasche) [reply runtime/memprof.c]: I see. So if you had two functions as I suggested above, but with the current approach of "morally setting lambda to 0", they could look like

```c
void caml_memprof_suspend_sampling(void) {
  if (caml_memprof_suspended) return;
  caml_memprof_suspended = 1;
  set_trigger(caml_young_alloc_start);
}

void caml_memprof_resume_sampling(void) {
  if (!caml_memprof_suspended) return;
  caml_memprof_suspended = 0;
  caml_memprof_renew_minor_sample();
}
```

I can see now why the two are equivalent and your approach is more factorized, but at first it was a bit confusing to see logic to renew the sample when *suspending* sampling (actually it does not renew but, on the contrary, it removes the trigger).

Another implementation option one could use it to save the current `lambda` value in a `caml_memprof_suspended_lambda`, and actually set `lambda` to `0`, instead of keeping a separate boolean variable. This makes the check-sites slightly simpler.

If I understand correctly, the idea of the second implementation you are talking about is to check for `caml_memprof_suspended` in the `caml_memprof_track_*` functions and return directly in that case. (That does look very simple.)

File: runtime/memprof.c
Comment (gasche) [reply runtime/memprof.c]: As you feel. It's probably fine as it is, but making the assumption explicit in my comment helped me review the code (and maybe it could help others).

I'm not familiar with the minor GC so it wasn't immediate for me to check that assumption (the wording of the documentation comment in `runtime/minor_gc.c` suggests it, but does not clearly say it. The place where it is made clear is in `runtime/caml/memory.h`, which is not the first place I thought of looking.

File: runtime/signals_nat.c
Comment (gasche) [inline runtime/signals_nat.c:81]: I have the impression that the current value of `caml_young_trigger` is lost with this change. Why is this a correct thing to do? One argument would be that the `renew_minor_sample` trigger will always fire before the existing one, but it's not clear to me that this is the case when `caml_young_trigger` is `caml_young_alloc_mid` and `caml_young_ptr` is above that (it looks like this can advance past `caml_young_alloc_mid`).

File: runtime/minor_gc.c
Comment (gasche) [inline runtime/minor_gc.c]: I think that I understand what this code does now. I wondered if it would be possible to always renew the sample here, and call `caml_memprof_track_young` after that -- if I understand correctly, you wouldn't need to renew at all in `caml_memprof_track_young` in that case. The reason why you are not doing this is that the beginning of `caml_memprof_track_young` needs access to the pre-renewal value of `caml_memprof_young_trigger`. Would it be nice to save this value here (or even the difference with the young pointer) and pass it as a parameter to `caml_memprof_track_young` instead?

File: runtime/memprof.c
Comment (jhjourdan) [reply runtime/memprof.c]: > If I understand correctly, the idea of the second implementation you are talking about is to check for caml_memprof_suspended in the caml_memprof_track_* functions and return directly in that case. (That does look very simple.)

Indeed. I pushed that new implementation.

File: runtime/signals_nat.c
Comment (jhjourdan) [reply runtime/signals_nat.c:81]: `caml_young_trigger` is not changed by `caml_memprof_renew_minor_sample();`, so I do not see what is the problem you are speaking about.

Are you speaking about `caml_memprof_young_trigger`? Then, yes, indeed, the value is lost. To see why this is correct, we have to distinguish two cases:
- Either the allocation that called `caml_garbage_collection` did not trigger the allocation. Then, yes, we will forget the position of the next sample, but, because of the memorylessness of the exponential distribution, the distribution of the next sample  will be the same.
- Either it triggered it. Temporarily, sampling in native code is not supported (in this PR). So, the only thing we can do is ignoring this sample. We do this by generating a new sample, and hope that it will not happen to re-sample the same block (otherwise we will just retry... no real harm here).

File: runtime/minor_gc.c
Comment (jhjourdan) [reply runtime/minor_gc.c]: `caml_memprof_young_trigger` is not the only variable to save. We would also have to save `next_sample_frac_part`. I can do what you propose (I do something similar for native sampling), but this essentially requires to split `caml_memprof_track_young` in two...

File: runtime/signals_nat.c
Comment (gasche) [reply runtime/signals_nat.c:81]: I was thinking of the write to `caml_young_trigger` being not-written and a trigger going missing, but indeed now I can see (again) that `set_trigger` does the right thing. Sorry for the noise.

File: runtime/minor_gc.c
Comment (gasche) [reply runtime/minor_gc.c]: I see. I think the current version is reasonable, so it's probably not worth changing, although the split version could be conceptually nicer -- it splits the responsability between the part that maintains the evolving state, and the part that consumes it.

File: runtime/memprof.c
Comment (gasche) [inline runtime/memprof.c]: if

File: runtime/memprof.c
Comment (gasche) [inline runtime/memprof.c]: Is this comment still relevant with the new `suspend` implementation? (Maybe it could be replaced or completed by a `CAMLassert`.)

File: runtime/memprof.c
Comment (gasche) [inline runtime/memprof.c:456]: Just out of curiosity: is it important to do this `caml_gc_dispatch` call before the rest of the memprof-specific logic below? If not, could this be turned instead into a recursive call to `caml_alloc_small_dispatch`? (Dispatching again after the callback call.)

**Review by gasche — COMMENTED** [review]
I have looked at this code and I believe it is correct -- although if I understand correctly @jhjourdan still has to make a change to `caml_alloc_small_dispatch` to call `caml_gc_dispatch` more often, along the lines mentioned in https://github.com/ocaml/ocaml/pull/8668/files#r283093142 .

This change is touching a performance-critical part of the compiler, but if I read the code correctly all changes happen in "cold" paths (when triggers trigger, which is infrequent and was already tested before the change), so there should be no performance regression when `lambda = 0`.

I didn't "approve" the PR because I think that an expert (for example @damiendoligez or @stedolan) should also look at this.

File: runtime/memprof.c
Comment (jhjourdan) [reply runtime/memprof.c:456]: Yes, this is important. Otherwise, we do not know the pointer to the block which is going to be sampled, and we cannot assign it to the ephemeron.

File: runtime/memprof.c
Comment (stedolan) [inline runtime/memprof.c]: Please find an alternative way to do this control flow!

File: runtime/memprof.c
Comment (stedolan) [inline runtime/memprof.c]: I did not realise until the second reading that this function was different from the similarly-named one in `minor_gc.c`. I think it would be clearer to update `caml_memprof_young_trigger` directly from this file (and update `caml_young_trigger` directly in `minor_gc.c`), and have a single function `caml_update_young_limit(void)` that sets `caml_young_limit` to the max of these two.

File: runtime/memprof.c
Comment (stedolan) [inline runtime/memprof.c:455]: The invariants here are subtle and undocumented and it took me a long time to work them out. I'd like this logic to be simplified a bit, or at least to have the invariants documented. In particular:

  - The word sampled is not the one that memprof_young_trigger points to (because capturing the callstack and running the callback allocates), but the one below memprof_young_trigger by a number of words equal to the allocation done by the callback. (This may be below young_limit, in which case a GC occurs).
 
  - This function is only called from `caml_alloc_small_dispatch` and `caml_young_ptr` is never modified between the time that this function returns and the time that `Alloc_small` completes. (Otherwise, the value sampled would not be the value allocated)

  - The functions `caml_ephemeron_set_key` (and `do_check_key_clean`, `do_set`, `caml_realloc_ephe_ref_table`, etc.) never inspect the new key other than to check that it is a young block. (Otherwise, they would crash upon finding that the key does not have a valid tag and contains uninitialised garbage).

The API is also a bit tricky: an arbitrary amount of user code runs between the ephemeron being created and its key being set (because finalisers, signal handlers and context switches can happen in caml_gc_dispatch above), and this computation can include further memprof sampling. So, in edge cases there can be several outstanding ephemerons with unset keys, and their keys can remain unset for a long time. In particular, polling the ephemeron to see if the sampled block is still alive does not always give the right answer.

**Review by stedolan — COMMENTED** [review]
I've been wondering something about statmemprof: why does it use the Poisson distribution instead of a binomial one?

A binomial distribution with parameter 0.01 would mean that, on average, 1% of all allocated words are sampled. A Poisson distribution with parameter 0.01 means that sampling events occur continuously at an average rate 1% the allocation rate (in particular, slightly less than 1% of all allocated words are sampled, with some sampled more than once). In other words, the binomial distribution is discrete-time and the Poisson is continuous.

In the previous PR, this distinction made little difference: at low parameter values, the two distributions are very close. However, this PR contains logic to handle the continuous-time Poisson process (the code dealing with `next_sample_frac_part`). Is there any advantage to having this? I find the binomial one easier to understand and explain, and we'd get to delete this code.

File: runtime/memprof.c
Comment (jhjourdan) [reply runtime/memprof.c]: Do you prefer a `goto`? Or code duplication?

**Comment by jhjourdan** [comment]
The reason we use Poisson distribution instead of a Binomial one is twofold:

- I usually have less trouble reasoning about continuous distribution rather than discrete ones. I.e., generating an exponential deviate is simpler than generating a geometric deviate (this involves computing trunc(log(U)/log(1-p)), if I am not mistaken, instead of the simpler log(U)/lambda).

- For large blocks, it is easier to generate a Poisson distribution than a binomial distribution using a rejection-based method such as the one used in `mt_generate_poisson` for large `cur_lambda`.

That said, I agree that the two distributions have very similar statistical properties and that they would be both valid for profiling. I also agree that the second argument above is not really valid anymore since we agreed that fast sampling of large blocks was not strictly necessary. In addition, there is another reason the binomial distribution could be better: when setting the sampling rate to 1, we actually sample each bock exactly once, which is not an easy property to get with the Poisson process.

For these reasons, let me try to propose to change the code for the binomial distribution, and let's see where does this get us.

File: runtime/memprof.c
Comment (stedolan) [reply runtime/memprof.c]: In general, yes: both `goto` and code duplication (especially when so short) are preferable to having nonlocal control flow in a preprocessor macro.

If you dislike those, the function can be written without either: (warning: uncompiled and untested)

```c
void caml_memprof_renew_minor_sample(void) {
  value* new_trigger = caml_young_alloc_start;

  if (lambda > 0) {
    double exp = - lambda_inverse * logf(mt_generate_uniform());
    uintnat max = caml_young_ptr - caml_young_alloc_start;
    if (exp < (double)max) {
      uintnat exp_int = (uintnat)exp;
      /* This assertion can be false only if [max] is not exactly
         representable as a double, which cannot happen if the size of the
         minor heap does not reach 2^52. */
      CAMLassert(exp_int < max);

      new_trigger = caml_young_ptr - exp_int;
      next_sample_frac_part = exp - exp_int;
      CAMLassert(0 <= next_sample_frac_part && next_sample_frac_part < 1);
    }
  }

  set_trigger(new_trigger);
}
```

I'm fine with other ways of writing this function too, but please avoid macros for control flow.

Comment (damiendoligez) [reply runtime/memprof.c]: I think a `goto` is the right thing here. Think of it as a kind of static exception and it won't be harmful.

File: runtime/memprof.c
Comment (damiendoligez) [inline runtime/memprof.c]: My eyes are bleeding each time I see the French spelling of "occurrence" in this PR. Could you switch to the English spelling?

File: runtime/caml/memory.h
Comment (damiendoligez) [reply runtime/caml/memory.h:220]: This change is indeed a bug fix.

File: runtime/memprof.c
Comment (jhjourdan) [reply runtime/memprof.c:455]: > The API is also a bit tricky: an arbitrary amount of user code runs between the ephemeron being created and its key being set (because finalisers, signal handlers and context switches can happen in caml_gc_dispatch above), and this computation can include further memprof sampling. So, in edge cases there can be several outstanding ephemerons with unset keys, and their keys can remain unset for a long time. In particular, polling the ephemeron to see if the sampled block is still alive does not always give the right answer.

Good catch! I've "fixed" the issue by filling the ephemeron with `Val_unit` during the `caml_gc_dispatch` call, so that it never appears empty before the block is freed. Since unboxed values like `Val_unit` are never freed by ephemerons, this also ensures that whatever the user put in the ephemeron's data will not be reclaimed.

Comment (jhjourdan) [reply runtime/memprof.c:455]: > The invariants here are subtle and undocumented and it took me a long time to work them out. I'd like this logic to be simplified a bit, or at least to have the invariants documented.

I tried to add comments. Better now?

I do agree that this code is rather subtle (this is why I prepared first the PR with major blocks only). That said, I don't see a good way of avoiding this complexity. One way would be to "postpone" sampling as we do for major blocks, but then, contrarily to major blocks, there is no good way of making sure the callback is indeed called shortly after....

**Comment by jhjourdan** [comment]
> For these reasons, let me try to propose to change the code for the binomial distribution, and let's see where does this get us.

I'll try to do that. But this is mostly independent from this PR (appart from the [next_sample_frac_part] business, but this is marginal). So let's first try to merge this, and then work on the binomial distribution.

**Comment by gasche** [comment]
Regarding comments: during my own review I did feel that comments on function declarations (or prototypes) on the pre/post-conditions of the functions would have been helpful, and possibly some mild effect typing: "{reads,consumes} and {refreshes,renews} the sampler state" could be a comment about `caml_memprof_track_young`, for example, as well as "is called just after a user demands a block, and creates values that refer to that block position in the heap; no OCaml code should run between the return and this function and the actual initialization of this demanded memory block".

(This implies some partial duplication of comments between the call-site and the declaration-site of a function. This is desirable because it makes understanding the code modular, without having to look at the declaration-site and the call-site at the same time, and will make it easier to understand the addition of new call sites in the future.)

I'm not saying that you (Jacques-Henri) should write more of these comments; I thought of writing some and sending them as a PR on top of your branch, but I can't do this before at least Saturday.

**Review by damiendoligez — APPROVED** [review]
I can't vouch for the probability stuff, but the GC-related changes look good to me.

**Comment by jhjourdan** [comment]
> I'm not saying that you (Jacques-Henri) should write more of these comments; I thought of writing some and sending them as a PR on top of your branch, but I can't do this before at least Saturday.

Ok, good enough. I tried my best to add comments, but it is actually difficult to figure out what is important to say and what is not when you wrote the code. Flooding the code with comments is not a good thing to do either...

**Comment by gasche** [comment]
Let's get this merged as soon as the current points have been addressed, no need to wait for my comments. @stedolan, are you also happy with the present state of the PR?

**Comment by mshinwell** [comment]
One stylistic remark: it would seem better to follow the convention elsewhere in the runtime of having curly braces delimiting the start of functions on their own line.  This is standard practice for C.

Regarding comments, I'm a bit concerned about duplication, since comments are prone to getting out of date anyway.  (I'd also prefer not to see comments that are just English versions of the code, rather than additional explanatory information, since those are particularly prone to going out of date.)

File: runtime/memprof.c
Comment (jhjourdan) [reply runtime/memprof.c]: AFAIK, this English and French spelling are identical:

https://www.wordreference.com/fren/occurrence

Comment (jhjourdan) [reply runtime/memprof.c]: Ah, but even then the spelling is incorrect. Fixed.

**Comment by jhjourdan** [comment]
I added a new commit for checking in `caml_update_young_limit` whether a signal is pending/a GC is requested. Therefore, a minor GC round will no longer hide signals.

This is a fix to a minor bug of the current trunk (signals can be delayed/ignored in native mode if minor GCs are triggered by something else than a native code allocation), but this could become more serious with statmemprof, since every minor sample could erase `caml_young_limit`.

**Comment by gasche** [comment]
Regarding comments and redundancy: I think that functions should document the assumptions that they make on the context around their call. In tricky cases, function calls should also document how/why those assumptions are satisfied. Whenever the latter happen, there is partial overlap, given that those assumptions are mentioned on both sides. I don't think that is a bad thing; in particular, if you change the assumptions made by a function implementation, then of course you should review all call-sites, and the potential need to update their documentation (~ correctness argument) is part of that process.

**Review by stedolan — APPROVED** [review]
I'm happy with the current version. Good catch with `uncaught_exception`, I missed that one.

> So let's first try to merge this, and then work on the binomial distribution.

I agree. I'm still interested in seeing if the binomial / geometric distribution improves things, but that's a seperate PR.

**Comment by jhjourdan** [comment]
> I'm happy with the current version. Good catch with `uncaught_exception`, I missed that one.
> 

Alright, I squashed the commits. The PR is ready to merge.

> > So let's first try to merge this, and then work on the binomial distribution.
> 
> I agree. I'm still interested in seeing if the binomial / geometric distribution improves things, but that's a seperate PR.

I'm already working on this. I will propose a new PR shortly after that one is merged.

**Comment by shindere** [comment]
this PR seems to cause a failure on Inria's CI:
(in the `tests/lib-dynlink-pr4839` tset):

```
 ... testing 'test.ml' with 1.2.2.17 (ocamlopt.byte) => /usr/bin/ld: Warning: size of symbol `caml_memprof_postponed_head' changed from 8 in /home/barsac/ci/builds/workspace/extra-checks/runtime/libasmrun.a(main_n.o) to 64 in /home/barsac/ci/builds/workspace/extra-checks/runtime/libasmrun.a(memprof_n.o)
passed
```

It's not a real failure because the test actually passes,
but it seems that the linker warning breaks the awk script that parses
the test's output.

FWIW this test is run with OCaml configured with:
```
./configure CC=clang-6.0
```

**Comment by jhjourdan** [comment]
Thanks, @gasche, for merging.

Thanks, @shindere for notifying me of the failure. I cannot reproduce it with my version of CLang (7.0.1) on Fedora, but I have a guess for fixing the issue: in `runtime/caml/memprof.h`, the global variable `caml_memprof_postponed_head` is not declared as `extern`, as it should.

Since this is anyway a defect in my code, and that fixing it could fix the warning, could you please try to push that change and see what happens?

**Comment by shindere** [comment]
Jacques-Henri Jourdan (2019/05/20 07:14 -0700):
> Thanks, @shindere for notifying me of the failure.

No problem. Just to make sure, are you subscribed to the
`ci-ocaml@inria.fr` mailing list? If not, it may make sense to do so.

> I cannot reproduce it with my version of CLang (7.0.1) on Fedora, but I have a guess for fixing the issue: in `runtime/caml/memprof.h`, the global variable `caml_memprof_postponed_head` is not declared as `extern`, as it should.
> 
> Since this is anyway a defect in my code, and that fixing it could fix
> the warning, could you please try to push that change and see what
> happens?

Done in commit a866412b2d2237ebbbae6fe72156be5f32d84b2e

**Comment by jhjourdan** [comment]
> No problem. Just to make sure, are you subscribed to the `ci-ocaml@inria.fr` mailing list? If not, it may make sense to do so.

I am not. Is this a public mailing list?

**Comment by shindere** [comment]
Jacques-Henri Jourdan (2019/05/20 07:39 -0700):
> > No problem. Just to make sure, are you subscribed to the `ci-ocaml@inria.fr` mailing list? If not, it may make sense to do so.
> 
> I am not. Is this a public mailing list?

I think so. You should check on sympa.inria.fr

**Pull request merged**
